node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java

```diff
@@ -139,20 +140,29 @@ private boolean deployTo(Move move) {
         try (MaintenanceDeployment deployment = new MaintenanceDeployment(application, deployer, nodeRepository())) {
             if ( ! deployment.isValid()) return false;
 
-            boolean couldMarkRetiredNow = markWantToRetire(move.node, true);
+            boolean couldMarkRetiredNow = markWantToRetire(Optional.of(move.node), true);
             if ( ! couldMarkRetiredNow) return false;
 
+            Optional<Node> expectedNewNode = Optional.empty();
             try {
                 if ( ! deployment.prepare()) return false;
-                if (nodeRepository().getNodes(application, Node.State.reserved).stream().noneMatch(node -> node.hasParent(move.toHost.hostname())))
-                    return false; // Deployment is not moving the from node to the target we identified for some reason
+                expectedNewNode =
+                        nodeRepository().getNodes(application, Node.State.reserved).stream()
+                                        .filter(node -> node.hasParent(move.toHost.hostname()))
+                                        .filter(node -> node.allocation().get().membership().cluster().id().equals(move.node.allocation().get().membership().cluster().id()))
+                                        .findAny();
+                if (expectedNewNode.isEmpty()) return false;
                 if ( ! deployment.activate()) return false;
 
                 log.info("Rebalancer redeployed " + application + " to " + move);
                 return true;
             }
             finally {
-                markWantToRetire(move.node, false); // Necessary if this failed, no-op otherwise
+                markWantToRetire(nodeRepository().getNode(move.node.hostname()), false); // Necessary if this failed, no-op otherwise
+                if (expectedNewNode.isPresent()) { // Immediately clean up if we reserved the node but could not activate
+                     Optional<Node> reservedNewNode = nodeRepository().getNode(expectedNewNode.get().hostname(), Node.State.reserved);
```

Comment:
Why do we need to get it again? We are still inside the lock

Reply:
Yes, but the deployment action can have changed it.